lib/MessageIdImpl.h

```diff
@@ -33,6 +33,8 @@ class MessageIdImpl {
           partition_(partition),
           batchIndex_(batchIndex),
           topicName_() {}
+    virtual ~MessageIdImpl() {}
```

Comment:
Yes. It's virtual only for adding RTTI. When there's no virtual method of the base class, it's impossible to use `dynamic_cast` (or `std::dynamic_pointer_cast` if it's a `std::shared_ptr`) because there is no RTTI info.

```c++
struct Base {
  int i = 1;
};
struct Derived : Base {
  int j = 2;
};

int main() {
  std::shared_ptr<Base> pb{new Derived};
  auto pd = std::dynamic_pointer_cast<Derived>(pb);
```

Compilation error:

> error: cannot dynamic_cast ‘(& __r)->std::shared_ptr<Base>::<anonymous>.std::__shared_ptr<Base, __gnu_cxx::_S_atomic>::get()’ (of type ‘using element_type = std::remove_extent<Base>::type*’ {aka ‘struct Base*’}) to type ‘using element_type = using element_type = std::remove_extent<Derived>::type*’ {aka ‘struct Derived*’} (source type is not polymorphic)

However, if we didn't have the precondition that `BatchMessageIdImpl` **is a** `MessageIdImpl`, there is no way to determine whether the `impl_` field of `MessageId` is a `BatchMessageIdImpl`.

The other reason is that make the destructor is the best practice when a class could be derived. See the following example:

```c++
#include <iostream>
#include <memory>
using namespace std;

struct Base {
  ~Base() { std::cout << "~Base" << std::endl; }
  virtual void f() {}
};
struct Derived : Base {
  ~Derived() { std::cout << "~Derived" << std::endl; }
};

int main() {
  std::shared_ptr<Base> pb;
  pb.reset(dynamic_cast<Base*>(new Derived));
  pb->f();
  return 0;
}
```

You will see the output:

```
~Base
```

It's because when the destructor of `Base` is called, the destructor of the derived class won't be called because the destructor is non-virtual and there is no RTTI. The code above is unnatural, but we still need to take carefully of the code to avoid such cases. The solution is to make the destructor of the base class virtual, so the destructor of the derived class could be called.

Reply:
Thanks for your detailed explanation. Get it.